src/plugins/hiddenColumns/contextMenuItem/showColumn.js

```diff
@@ -6,112 +6,100 @@ import * as C from '../../../i18n/constants';
     },
     callback() {
-      const { from, to } = this.getSelectedRangeLast();
-      const start = Math.min(from.col, to.col);
-      const end = Math.max(from.col, to.col);
-
-      if (start === end) {
-        if (beforeHiddenColumns.length === start) {
-          hiddenColumnsPlugin.showColumns(beforeHiddenColumns);
-          beforeHiddenColumns.length = 0;
-        }
-        if (afterHiddenColumns.length === this.countSourceCols() - (start + 1)) {
-          hiddenColumnsPlugin.showColumns(afterHiddenColumns);
-          afterHiddenColumns.length = 0;
-        }
+      const [, startVisualColumn, , endVisualColumn] = this.getSelectedLast();
+      const onlyFirstVisibleColumnSelected =
+        this.columnIndexMapper.getRenderableFromVisualIndex(startVisualColumn) === 0 &&
+        startVisualColumn === endVisualColumn;
+      const onlyLastVisibleColumnSelected =
+        this.columnIndexMapper.getRenderableFromVisualIndex(endVisualColumn) === this.countRenderableColumns() - 1 &&
+        startVisualColumn === endVisualColumn;
+
+      let startPhysicalColumn = this.toPhysicalColumn(startVisualColumn);
+      let endPhysicalColumn = this.toPhysicalColumn(endVisualColumn);
+
+      if (onlyFirstVisibleColumnSelected) {
+        startPhysicalColumn = 0;
+      }
 
-      } else {
-        rangeEach(start, end, column => hiddenColumnsPlugin.showColumn(column));
+      if (onlyLastVisibleColumnSelected) {
+        endPhysicalColumn = this.countSourceCols() - 1;
       }
 
+      hiddenColumnsPlugin.showColumns(columns);
+
+      const startVisualColumnAfterAction = this.toVisualColumn(startPhysicalColumn);
+      const endVisualColumnAfterAction = this.toVisualColumn(endPhysicalColumn);
+
+      // Selection start and selection end coordinates might be changed after showing some items.
+      this.selectColumns(startVisualColumnAfterAction, endVisualColumnAfterAction);
       this.render();
       this.view.wt.wtOverlays.adjustElementsSize(true);
     },
     disabled: false,
     hidden() {
-      if (!hiddenColumnsPlugin.hiddenColumns.length || !this.selection.isSelectedByColumnHeader()) {
+      if (!this.selection.isSelectedByColumnHeader() || hiddenColumnsPlugin.getHiddenColumns().length < 1) {
         return true;
       }
 
-      beforeHiddenColumns.length = 0;
-      afterHiddenColumns.length = 0;
+      columns.length = 0;
 
-      const { from, to } = this.getSelectedRangeLast();
-      const start = Math.min(from.col, to.col);
-      const end = Math.max(from.col, to.col);
-      let hiddenInSelection = false;
+      const [, startColumn, , endColumn] = this.getSelectedLast();
+      const start = Math.min(startColumn, endColumn);
+      const end = Math.max(startColumn, endColumn);
+      const sequence = this.columnIndexMapper.getIndexesSequence();
+      const physicalStartColumn = this.columnIndexMapper.getPhysicalFromVisualIndex(start);
+      const currentStartPosition = sequence.indexOf(physicalStartColumn);
```

Comment:
I think this is another spot where I see that we need to check `indexOf` on the columnMapper sequence. Maybe we're missing some API on the IndexMapper?

Reply:
As above.